lib/geoengineer/gps/gps.rb

```diff
@@ -0,0 +1,305 @@
+      # if the partial file exists we load the file directly
+      # This will create the GPS resources
+      require "#{Dir.pwd}/#{partial_file_name}"
+    else
+      # otherwise initalize for the partial directly here
+      gps_instance.partial_of(partial_file_name)
+    end
+  end
+
+  def load_gps_file(gps_file)
+    GeoEngineer::GPS.load_gps_file(self, gps_file)
+  end
+
+  # Parse
+  def self.parse_dir(dir)
+    base_hash = {}
+
+    extension = ".gps.yml"
+
+    # Load, expand then merge all yml files
+    Dir["#{dir}**/*#{extension}"].each do |gps_file|
+      # Merge Keys don't work with YAML.safe_load
+      # since we are also loading Ruby safe_load is not needed
+
+      gps_text = ERB.new(File.read(gps_file)).result(binding).to_s
+      gps_hash = YAML.load(gps_text)
+      # remove all keys starting with `_` to remove paritals
+      gps_hash = remove_(gps_hash)
+      JSON::Validator.validate!(json_schema, gps_hash)
+
+      # project name it the path + file
+      project_name = gps_file.sub(dir, "")[0...-extension.length]
+
+      # assign to the base_hash the
+      base_hash[project_name.to_s] = gps_hash
+    end
+
+    GeoEngineer::GPS.new(base_hash)
+  end
+
+  attr_reader :nodes
+  def initialize(base_hash)
+    # Base Hash is the unedited input, useful for debugging
+    @base_hash = base_hash
+
+    # First Deep Dup to ensure seperate objects
+    # Dup to ensure string keys and to expeand
+    projects_hash = GeoEngineer::GPS.deep_dup(base_hash)
+
+    # expand meta nodes, this takes nodes and expands them
+    projects_hash = expand_meta_nodes(projects_hash)
+
+    # build the node instances and add them to all nodes
+    @nodes = build_nodes(projects_hash)
+
+    # validate all nodes
+    @nodes.each(&:validate) # this will validate and expand based on their json schema
+  end
+
+  def find(query)
+    GeoEngineer::GPS.find(@nodes, query)
+  end
+
+  def where(query)
+    GeoEngineer::GPS.where(@nodes, query)
+  end
+
+  def to_h
+    GeoEngineer::GPS.deep_dup(@base_hash)
+  end
+
+  def expanded_hash
+    expanded_hash = {}
+    @nodes.each do |n|
+      proj = expanded_hash[n.project] ||= {}
+      env = proj[n.environment] ||= {}
+      conf = env[n.configuration] ||= {}
+      nt = conf[n.node_type] ||= {}
+      nt[n.node_name] ||= n.attributes
+    end
+    expanded_hash
+  end
+
+  def loop_projects_hash(projects_hash)
+    # TODO: validate the strucutre before this
+    projects_hash.each_pair do |project, environments|
+      environments.each_pair do |environment, configurations|
+        configurations.each_pair do |configuration, nodes|
+          nodes.each_pair do |node_type, node_names|
+            node_names.each_pair do |node_name, attributes|
+              node_type_class = find_node_class(node_type)
+              node = node_type_class.new(project, environment, configuration, node_name, attributes)
+              yield environments, configurations, nodes, node_names, node
```

Comment:
do we ever use anything other than the node?

Reply:
Seems no - maybe reverse the order so we don't have `|_, _, _, _, node|` - looks a bit ugly

Reply:
Not too fussed on this as it is only an internal method.